File: app/models/track.rb
require 'json'
require 'rubygems'
require 'yaml'
require 'pp'

module Nerve; module Model

	class Track < ActiveRecord::Base

		belongs_to :artist, foreign_key: 'artist'
		belongs_to :album, foreign_key: 'album'

		belongs_to :created_by, :class_name => "User", :foreign_key => 'created_by'
		belongs_to :approved_by, :class_name => "User", :foreign_key => 'approved_by'

		alias_attribute :upload_date, :creation_date
		alias_attribute :genre, :main_genre

		serialize :extra, JSON

		default_scope { includes :artist, :album }

		def approved
			!!status and status > 3
		end

		def update_metadata

			result = Nerve::Services::Metadata.match artist.name, (album.name rescue ''), title, true, true
			external_id = result['external_id']
			get_metadata 

		end

		def get_metadata

			result, cache_item = Nerve::Services::Metadata.match_meta(external_id, true, "external_id", true)
			_big = result["big"] rescue nil

			_lyrics = (cache_item.lyrics or JSON.parse(_big["lyrics"])[0]) rescue ""
			_lyrics = _lyrics[0] if _lyrics.is_a? Array
			_lyrics = '' if !_lyrics or _lyrics.to_s.empty? or _lyrics.to_s == '0' or _lyrics.to_s == 'false'

			if _lyrics == ''
				_lyrics = "No lyrics available."
				begin
					_lyrics = Nerve::Services::Metadata.match_lyrics(artist.name, (album.name rescue ''), title)[0] 
					if cache_item
						cache_item.lyrics = _lyrics
						cache_item.save
					end
				rescue
					_lyrics = "No lyrics available."
				end
			end

			[result, _lyrics, (result["year"] rescue nil)]

		end

		# Unguaranteed to be absolutely safe, but it's more safe than sorry.
		def is_safe

			return false if explicit or status == 0
			_extended, _lyrics = get_metadata

			return true if instrumental

			# This nasty looking regex basically matches all words that start with nasties. 
			# that way we don't match, say, saltwater (although we might actually want to, it's gross)
			r = Regexp.new("\\b((#{$config["words"]["banned"].join("|")})[^\\s\\b,.\<\>]*)\\b", 7)

			# If the size is 0, then it's "safe" ish
			return false if !_lyrics or _lyrics.to_s.empty? or _lyrics == 0

			# TODO: make this cleaner
			_lyrics.scan(r).size == 0 && !!_lyrics && _lyrics != "" && _lyrics != "No lyrics available."

		end

		# cut point after extro is 8 seconds
		def playout_length
			[length, outro + 8.0].min
		end

		def set_unsafe
			explicit = true
		end

		def why_unsafe

			reasons = []
			reasons << "determined as risky during upload" if status == 0 # only works before submission
			reasons << "has parental advisory/explicit flag" if explicit
			_extended, _lyrics = get_metadata

			if !_lyrics.to_s.empty? && _lyrics != 0 && _lyrics != "No lyrics available."
				r = Regexp.new("\\b((#{$config["words"]["banned"].join("|")})[^\\s\\b,.\<\>]*)\\b", 7)
				s = _lyrics.scan(r)
				reasons << "it contains (at least one) expletive (#{s[0]})" if s.size > 0
			else
				reasons << "no lyrics were found"
			end

			reasons.length > 0 ? reasons : false

		end

		def delete! soft = false

			_base = $config["export"]["directory"] + "/"
			File.unlink(_base + local_path) rescue puts "Failed to delete #{id}: #{$!}"

			if soft
				self.status = 6
				self.save
			else
				File.unlink(_base + local_path_preview) rescue puts "Failed to delete #{id} preview: #{$!}"
				File.unlink(_base + local_path_waveform) rescue puts "Failed to delete #{id} form: #{$!}"
				destroy
			end

		end

		def get_json extended = false

			user = created_by

			data = {
				"id" => id,
				"external_id" => external_id, 
				"title" => title,
				"artist" => artist.name,
				"album" => album.name,
				"approved" => approved,
				"approved_by" => approved_by && approved_by.get_json,
				"genre" => genre,
				"genre_text" => $genres[genre],
				"length" => length,
				"status" => status,
				"upload_date" => upload_date.iso8601,
				"extro_start" => outro,
				"is_library" => is_library,
				"is_automation" => is_automation,
				"playout_id" => playout_id,
				"end_type" => end_type,
				"created_by" => user && user.get_json,
				"flagged" => flagged,
				"instrumental" => instrumental,
				"extra" => extra
			}

			# Extended means lyrics and such 
			if extended == true
				extended, lyrics = get_metadata
				data.merge!({
					"intro_start" => intro_start,
					"intro_end" => intro_end,
					"hook_start" => hook_start,
					"hook_end" => hook_end,
					"extro_start" => outro	})
			else
				extended = false
			end
			
			data["big"] = extended
			data["lyrics"] = lyrics

			data

		end

		def to_json extended = true

			self.get_json(extended).to_json

		end

	end

end; end

File: public/src/app/struct/track.ts
var dateFormat = require('dateformat');
var mm = require('../../../musicmetadata');

import { NerveService } from '../nerve.service';

export class Track {

	id: number;
	external_id: number;
	title: string;
	artist: string;
	album: string;

	approved: boolean;
	approved_by: number;
	genre: number;
	genre_text: string;

	length: number;
	status: number;
	upload_date: Date;

	is_library: boolean;
	is_automation: boolean;

	playout_id: string;
	end_type: number;

	// created_by: User;
	created_by: any;

	flagged: boolean;
	instrumental: boolean;

	ignore:string[] = ['big', 'genre_text', 'approved_by', 'approved', 'created_by', 'ignore', 'playout_id', 'status'];

	constructor (data: any) {

		for (var d in data) {

			switch (d) {
				case 'upload_date': data[d] = new Date(Date.parse(data[d])); break;
				// case 'upload_date': data[d] = new Date(d); break;
			}

			this[d] = data[d];

		}

	}

	serialize () {

		var keys = Object.keys(this);
		var data = {};

		for(var key in keys) {
			key = keys[key];
			if (this.ignore && this.ignore.indexOf(key) !== -1) {
				continue;
			}
			if (this[key] != null && (typeof this[key] == 'function' || typeof this[key] == 'object')) {
				console.log(typeof this[key], this[key], '!')
				if (typeof this[key]['serialize'] != 'undefined')
					data[key] = this[key].serialize()
				continue;
			}
			data[key] = this[key];
		}

		return data;

	}

	getUploadDate () {
		return dateFormat(this.upload_date, "dS mmmm yyyy h:MM:ss");
	}

	get preview_url () : string {
		return '/api/audio/preview/' + this.id;
	}

	get wave_url () : string {
		return '/api/audio/waveform/' + this.id;
	}

	get audio_url () : string|boolean {
		return this.status != 6 ? '/api/audio/get/' + this.id : false;
	}

	get submittable () {
		return this.id > 0 && this.end_type > 0 && this.extro_start > 0;
	}

}

export class UploadTrack extends Track {

	static from (file:File, metadata?: Object) : Promise<UploadTrack> {

		return new Promise<UploadTrack>(function(fulfil, reject) {

			if (metadata) {
				return fulfil(metadata);
			}

			let extension = file.name.split(".").pop().toLowerCase()

			// We should handle this further up in executon
			if (extension == "wav") {
				reject(['metadata']);
			}

			var t = setTimeout(() => reject(['timeout']), 5000);
			var metadata = mm(file, function (err, metadata) {
				clearTimeout(t);
				if (err) {
					console.log('argh')
					return reject(['metadata', err]);
				}
				metadata.artist = metadata.artist[0];
				fulfil(metadata)
			})

		}).then(UploadTrack.search);

	}

	private static search (metadata: Object) {
		return NerveService.getInstance().metadataMatch(metadata);
	}

}

export class IterTrack extends Track {

	get click () : boolean {

		return this.status < 2;

	}

	get greyed () : boolean {
		return !this.click;
	}

}

export class FullTrack extends Track {

	intro_start: number;
	intro_end: number;
	hook_start: number;
	hook_end: number;
	extro_start: number;

	end_type:number;

	big: any;

}